proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,317 @@
+
+```kotlin
+private val _items = mutableListOf<Item>()
+val item : List<Item> by _items
+```
+
+And the new syntax allows us to write:
+
+```kotlin
+val items: List<Item>
+	private field = mutableListOf()
+```
+
+### Android Architecture Components
+
+>  Proper way to do Architecture components is to use `MutableLiveData` (`LiveData` implementation that allows caller to change its value) privately inside View Model classes and then only expose read-only `LiveData` objects outside.
+
+Sample code [from an Android app](https://github.com/elpassion/crweather/blob/9c3e3cb803b7e4fffbb010ff085ac56645c9774d/app/src/main/java/com/elpassion/crweather/MainModel.kt#L14):
+
+```kotlin
+private val mutableCity = MutableLiveData<String>().apply { value = "" }
+private val mutableCharts = MutableLiveData<List<Chart>>().apply { value = emptyList() }
+private val mutableLoading = MutableLiveData<Boolean>().apply { value = false }
+private val mutableMessage = MutableLiveData<String>()
+
+val city: LiveData<String> = mutableCity
+val charts: LiveData<List<Chart>> = mutableCharts
+val loading: LiveData<Boolean> = mutableLoading
+val message: LiveData<String> = mutableMessage
+```
+
+Becomes:
+
+```kotlin
+val city: LiveData<String>
+	field = MutableLiveData().apply { value = "" }
+val charts: LiveData<List<Chart>>
+	field = MutableLiveData().apply { value = emptyList() }
+val loading: LiveData<Boolean>
+	field = MutableLiveData().apply { value = false }
+val message: LiveData<String>
+	field = MutableLiveData()
+```
+
+### RX Observable and Subjects
+
+> Data can often be pushed into reactive world using subjects. However, exposing `Subject` would allow consumer of the class to push its own data into it. That is why it is good idea to expose all subjects as read-only `Observable`:
+
+```kotlin
+class MyClass {
+    private val _dataStream = PublishSubject.create<String>()
+    val dataStream: Observable<String>
+        get() = _dataStream
+}
+```
+
+Turns into:
+
+```kotlin
+class MyClass {
+    val dataStream: Observable<String>
+    	field = PublishSubject.create()
+}
+```
+
+## Design
+
+The proposed design consists of two new ideas.
+
+### Explicit Backing Fields
+
+```kotlin
+val it: A
+	[visibility] field[: B] = initializer
+```
+
+The above `field` declaration is referred to as an _explicit backing field declaration_.
+
+#### Accessors
+
+- if `A :> B`, the compiler can provide a default getter
+- if `A <: B`, the compiler can provide a default setter
+- If an accessor is required, it must be provided either by the compiler or the user
+
+```kotlin
+public val flow: SharedFlow
+    field: MutableSharedFlow? = null
+    get() {
+        return field ?: run { ... }
+    }
+```
+
+#### Visibility
+
+Only the `private` and the `internal` visibilities are allowed for explicit backing fields now. The default visibility is `private`.
```

Comment:
May be mention reasoning why don't we want `protected` one.

Reply:
The thing is, we put `protected` away simply to let ourselves concentrate on a smaller task. Since STN is forbidden for non-`final` properties, there should be no ambiguity as to which backing field to use.

Reply:
Yes. Also, for Kotlin/Native protected fields might be a challenge to implement.